### PR TITLE
ART-3471, added new services

### PR DESCRIPTION
**Pull request by aleksei-artemev**

### CODE REVIEW

File: artacle-swagger.yaml [hunk 1]
```diff
@@ -314,6 +314,281 @@ paths:
...
+              - valuationRecent
+              - floor
+              - totalSalesVolume
+              - totalAmountOfSales
+            default: valuationAllTime
+            example: valuationAllTime
+        - name: sortingDirection
+          in: query
+          required: false
+          description: "Sorting direction: ascending, descending"
```
Comment (agamanin) [inline artacle-swagger.yaml]: Better use consistent values: asc, desc

File: artacle-swagger.yaml [hunk 2]
```diff
@@ -314,6 +314,281 @@ paths:
...
+          in: query
+          required: false
+          description: Query only traits with certain name
+          schema:
+            type: string
+            example: Color
+        - name: sortBy
+          in: query
+          required: false 
+          description: Sort result by one of criteria
```
Comment (agamanin) [inline artacle-swagger.yaml]: Minor - suggesting 'Traits sorting order in response'

File: artacle-swagger.yaml [hunk 3]
```diff
@@ -314,6 +314,281 @@ paths:
...
+          required: false
+          description: "Sorting direction: ascending, descending"
+          schema:
+            type: string
+            enum:
+              - asc
+              - desc
+            default: desc
+            example: desc
+        - name: offset
```
Comment (agamanin) [inline artacle-swagger.yaml:375]: Please, let's use all consistent descriptions for offset, limit and apiKey, as it used to be in other methods:
image

File: artacle-swagger.yaml [hunk 4]
```diff
@@ -314,6 +314,281 @@ paths:
...
+                type: object
+                required:
+                  - data
+                properties:
+                  data:
+                    type: array
+                    description: Array of Traits info
+                    items:
+                      $ref: '#/components/schemas/CollectionTraits'
+        '401':
```
Comment (agamanin) [inline artacle-swagger.yaml:409]: 400 should be added - Bad Request + see my comment about it for the source code

File: artacle-swagger.yaml [hunk 5]
```diff
@@ -314,6 +314,281 @@ paths:
...
+          required: false
+          description: Limit elements to fetch
+          schema:
+            type: integer
+            example: 50
+            default: 50
+            maximum: 1000
+      responses:
+        '200':
+          description: Success, returns array of collection traits, its valuation and information.
```
Comment (agamanin) [inline artacle-swagger.yaml]: Minor, THEIR valuation

File: artacle-swagger.yaml [hunk 6]
```diff
@@ -314,6 +314,281 @@ paths:
...
+        - name: apiKey
+          in: query
+          required: true
+          description: API access key
+          schema:
+            type: string
+        - name: traitName
+          in: query
+          required: true
+          description: Query only traits with certain name
```
Comment (agamanin) [inline artacle-swagger.yaml]: It is never 'traitS', it is always 1 trait, the description confuses. Suggesting: "Name of the trait", "Value of the trait"

File: artacle-swagger.yaml [hunk 7]
```diff
@@ -314,6 +314,281 @@ paths:
...
+          schema:
+            type: string
+            example: Color
+        - name: traitValue
+          in: query
+          required: true
+          description: Query only traits with certain value
+          schema:
+            type: string
+            example: Color
```
Comment (agamanin) [inline artacle-swagger.yaml]: Example "Black"

File: artacle-swagger.yaml [hunk 8]
```diff
@@ -314,6 +314,281 @@ paths:
...
+          in: query
+          required: true
+          description: Query only traits with certain name
+          schema:
+            type: string
+            example: Color
+        - name: traitValue
+          in: query
+          required: true
+          description: Query only traits with certain value
```
Comment (agamanin) [inline artacle-swagger.yaml]: Same as above

File: artacle-swagger.yaml [hunk 9]
```diff
@@ -314,6 +314,281 @@ paths:
...
+            application/json:
+              schema:
+                type: object
+                required:
+                  - data
+                properties:
+                  data:
+                    $ref: '#/components/schemas/CollectionTrait'
+
+        '401':
```
Comment (agamanin) [inline artacle-swagger.yaml:469]: 400 to be added

File: artacle-swagger.yaml [hunk 10]
```diff
@@ -314,6 +314,281 @@ paths:
...
+        - name: apiKey
+          in: query
+          required: true
+          description: API access key
+          schema:
+            type: string
+        - name: traitName
+          in: query
+          required: true
+          description: Query only traits with certain name
```
Comment (agamanin) [inline artacle-swagger.yaml]: Same as in the previous method

File: artacle-swagger.yaml [hunk 11]
```diff
@@ -314,6 +314,281 @@ paths:
...
+          schema:
+            type: string
+            example: Color
+        - name: traitValue
+          in: query
+          required: true
+          description: Query only traits with certain value
+          schema:
+            type: string
+            example: Color
```
Comment (agamanin) [inline artacle-swagger.yaml]: example: Black

File: artacle-swagger.yaml [hunk 12]
```diff
@@ -314,6 +314,281 @@ paths:
...
+        - name: startTimestamp
+          in: query
+          required: false
+          description: Unix timestamp. Indicates start time of collected samples. Method fetches everything till endTimestamp if not specified.
```
Comment (agamanin) [inline artacle-swagger.yaml]: 'collected samples' is quite vague. Suggesting 'sales to account in TTF'.

File: artacle-swagger.yaml [hunk 13]
```diff
@@ -314,6 +314,281 @@ paths:
...
+          schema:
+            type: number
+            format: uint64
+        - name: endTimestamp
+          in: query
+          required: false
+          description: Unix timestamp. Indicates end time of collected samples. Set to now by default.
```
Comment (agamanin) [inline artacle-swagger.yaml]: Same as above

File: artacle-swagger.yaml [hunk 14]
```diff
@@ -314,6 +314,281 @@ paths:
...
+          schema:
+            type: number
+            format: uint64
+        - name: includeSales
+          in: query
+          required: false
+          schema:
+            type: number
```
Comment (agamanin) [inline artacle-swagger.yaml:533]: I believe there must be native boolean support for it, is there?

File: artacle-swagger.yaml [hunk 15]
```diff
@@ -314,6 +314,281 @@ paths:
...
+          required: false
+          description: Limit elements to fetch
+          schema:
+            type: integer
+            example: 50
+            default: 50
+            maximum: 1000
+      responses:
+        '200':
+          description: Success, returns specific trait TTF information
```
Comment (agamanin) [inline artacle-swagger.yaml]: Suggesting to add a note 'within specified time range' - it is an important distinction of this method

File: artacle-swagger.yaml [hunk 16]
```diff
@@ -314,6 +314,281 @@ paths:
...
+                        type: number
+                        format: float
+                      sales:
+                        description: Historical information array. Included if includeSales is set to 1 in request
+                        type: array
+                        items:
+                          allOf:
+                          - $ref: '#/components/schemas/CollectionTraitSale'
+
+        '401':
```
Comment (agamanin) [inline artacle-swagger.yaml:587]: 400 code is needed

File: artacle-swagger.yaml [hunk 17]
```diff
@@ -918,6 +1194,202 @@ components:
...
+        - traitValue
+        - valuationAllTime
+        - valuationAllTimeUSD
+        - valuationRecent
+        - valuationRecentUSD
+        - recentPeriodMonths
+        - totalAmountOfSales
+        - medianTTFAllTime
+        - medianTTFRecentPeriod
+        - quantity
```
Comment (agamanin) [inline artacle-swagger.yaml]: Quantity must be renamed

File: artacle-swagger.yaml [hunk 18]
```diff
@@ -918,6 +1194,202 @@ components:
               nullable: true
               example: "0x9df947cd8a9810964473339c38a95201ecbe112d"
 
+    CollectionTraits:
+      type: object
+      required:
+        - traitValue
```
Comment (agamanin) [inline artacle-swagger.yaml:1207]: Floor is missing

File: artacle-swagger.yaml [hunk 19]
```diff
@@ -918,6 +1194,202 @@ components:
...
+        - valuationRecent
+        - valuationRecentUSD
+        - recentPeriodMonths
+        - totalAmountOfSales
+        - medianTTFAllTime
+        - medianTTFRecentPeriod
+        - quantity
+        - largeDistinctCount
+      properties:
+        traitValue:
```
Comment (agamanin) [inline artacle-swagger.yaml:1226]: We do use inconsistent notation here: we accept name and value separately in the input, and concatenate them in the output. I think that the output should provide name and value separately

File: artacle-swagger.yaml [hunk 20]
```diff
@@ -918,6 +1194,202 @@ components:
...
+      required:
+        - valuationAllTime
+        - valuationAllTimeUSD
+        - valuationRecent
+        - valuationRecentUSD
+        - recentPeriodMonths
+        - totalAmountOfSales
+        - medianTTFAllTime
+        - medianTTFRecentPeriod
+        - quantity
```
Comment (agamanin) [inline artacle-swagger.yaml]: Quantity must be renamed; floor (+USD) should be added

File: artacle-swagger.yaml [hunk 21]
```diff
@@ -918,6 +1194,202 @@ components:
...
+        totalSalesVolume:
+          description: Total sales volume
+          type: number
+          format: float
+        totalSalesVolumeUSD:
+          description: Total sales volume
+          type: number
+          format: float
+                
+    CollectionTrait:
```
Comment (agamanin) [inline artacle-swagger.yaml]: What is the difference with CollectionTraits? It looks like only CollectionTraiT must be kept and referred to in the methods

File: artacle-swagger.yaml [hunk 22]
```diff
@@ -918,6 +1194,202 @@ components:
...
+          type: number
+          format: float
+                
+    CollectionTraitSale:
+      type: object
+      required:
+        - timestamp
+        - ttfValue
+        - sale
+        - floorPrice
```
Comment (agamanin) [inline artacle-swagger.yaml:1289]: floorUSD is missing

File: artacle-swagger.yaml [hunk 23]
```diff
@@ -918,6 +1194,202 @@ components:
...
+        - salePrice
```
Comment (agamanin) [inline artacle-swagger.yaml:1291]: Price USD is missing

File: artacle-swagger.yaml [hunk 24]
```diff
@@ -918,6 +1194,202 @@ components:
...
+          description: Trait floor
+          type: number
+          format: float
+        salePriceUSD:
+          description: Trait floor
+          type: number
+          format: float
+    
+    # used in traits related methods
+    AmountAndCurrency:
```
Comment (agamanin) [inline artacle-swagger.yaml]: Not used anymore, can be removed

**Review by agamanin — CHANGES_REQUESTED** [review]

File: artacle-swagger.yaml [hunk 25]
```diff
@@ -314,6 +314,281 @@ paths:
         '500':
           $ref: "#/components/responses/InternalServerError"
 
+#  Collection Traits
```
Comment (agamanin) [inline artacle-swagger.yaml:322]: General: as requested by Alex, let's specifiy somewhere (we can do it in Summary of each method or somehow before the whole Traits block): "Trait's TTF (Trait (sale) to Floor ratio) is calculated as the ratio of prices of the last sale of each token having specific trait to the collection's floor price at the moment of sale. Outliers get filtered out, median is used to provide aggregated TTF characteristic that can estimate the trait's value"

**Review by agamanin — CHANGES_REQUESTED** [review]

File: artacle-swagger.yaml (hunk 1 above)
Comment (aleksei-artemev) [reply artacle-swagger.yaml]: ok, fixed

File: artacle-swagger.yaml (hunk 2 above)
Comment (aleksei-artemev) [reply artacle-swagger.yaml]: sounds good

File: artacle-swagger.yaml (hunk 3 above)
Comment (aleksei-artemev) [reply artacle-swagger.yaml:375]: fixed

File: artacle-swagger.yaml (hunk 5 above)
Comment (aleksei-artemev) [reply artacle-swagger.yaml]: fixed

File: artacle-swagger.yaml (hunk 20 above)
Comment (aleksei-artemev) [reply artacle-swagger.yaml]: fixed

File: artacle-swagger.yaml (hunk 22 above)
Comment (aleksei-artemev) [reply artacle-swagger.yaml:1289]: fixed

File: artacle-swagger.yaml (hunk 24 above)
Comment (aleksei-artemev) [reply artacle-swagger.yaml]: removed

File: artacle-swagger.yaml (hunk 25 above)
Comment (aleksei-artemev) [reply artacle-swagger.yaml:322]: added as tag description

**Pull request merged**